src/main/java/uk/gov/hmcts/dts/fact/mapit/MapitValidator.java

```diff
@@ -0,0 +1,110 @@
+import uk.gov.hmcts.dts.fact.services.MapitService;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Slf4j
+@Component
+public class MapitValidator {
+
+    private final MapitService mapitService;
+    private static final String FULL_POSTCODE = "([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z]"
+        + "[A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y]"
+        + "[0-9][A-Za-z]?))))\\s?[0-9][A-Za-z]{2})";
+    private static final String PARTIAL_POSTCODE_SEARCH = "[A-Za-z]{2}";
+    private static final String PARTIAL_POSTCODE_NUMERIC = "(^[A-Za-z]{1,2}([0-9]{1,3}))$";
+    private static final String PARTIAL_POSTCODE_EDGECASE = "(^[A-Za-z][0-9]{1,2}[A-Za-z][0-9]?$)|"
+        + "(^[A-Za-z]{2}([0-9]{1}[A-Za-z][0-9]?)$)";
+
+    @Autowired
+    public MapitValidator(MapitService mapitService) {
+        this.mapitService = mapitService;
+    }
+
+    public boolean postcodeDataExists(final String postcode) {
+
+        // Base validation for postcode
+        if (postcode.length() < 2 || StringUtils.isEmpty(postcode)) {
+            return false;
+        }
+
+        // If we have a full postcode
+        boolean fullResultFound = postcode.matches(FULL_POSTCODE) && fullPostCodeDataExists(postcode);
+
+        if (fullResultFound) {
+            log.info("Full postcode search result was successful for: {}", postcode);
+            return true;
+        }
+
+        // Partial postcode search criteria:
+        // One or two letters at the start, two or three digits
+        // If the starting is one letter:
+        //  - Sub part can be one number or two numbers or three numbers
+        //  - Or sub part can be one number, and one letter
+        //  - Or two numbers with one letter
+        //  - Or sub part can be one number, and one letter, and one number
+        //  - Or two numbers with one letter, and one number
+        // If the starting is two letters:
+        //  - sub part can be one number or two numbers or three numbers
+        //  - Or sub part can be one number, and one letter
+        //  - Or sub part can be one number, and one letter, and one number
+        return partialPostcodeValid(PARTIAL_POSTCODE_NUMERIC, postcode, 1)
+            // Check first two characters of the postcode, which will determine the result group we need
+            || (postcode.substring(0, 2).matches(PARTIAL_POSTCODE_SEARCH)
+                ? partialPostcodeValid(PARTIAL_POSTCODE_EDGECASE, postcode, 2) : // if two letter proceed
+                partialPostcodeValid(PARTIAL_POSTCODE_EDGECASE, postcode, 1));   // if one letters proceeds
+    }
+
+    private boolean fullPostCodeDataExists(String postcode) {
+        log.info("Full postcode search of mapit data for: {}", postcode);
+        boolean mapitDataExists = mapitService.getMapitData(postcode).isPresent();
+
+        if (mapitDataExists) {
+            return true;
+        } else {
```

Comment:
Don't need an else here as there is a return within the preceding if clause

Reply:
Removed the else, I want to keep the log.info in there though for debugging :)